### PR TITLE
debug: enable js-debug to auto attach

### LINKED ISSUE #88599: Adapt "auto attach" feature to js-debug
I would expect that using the existing "auto attach" feature of node-debug applies to js-debug too.

### PR DESCRIPTION
**Pull request by connor4312**
This modifies the debug-auto-launch extension to trigger js-debug as
outlined in https://github.com/microsoft/vscode/issues/88599#issuecomment-617242405

Since we now have four states, I moved the previous combinational logic
to a `transitions` map, which is more clear and reliable. The state
changes are also now a queue (in the form of a promise chain) which
should avoid race conditions.

There's some subtlety around how we cached the "ipcAddress" and know
that environment variables are set. The core desire is being able to
send a command to js-debug to set the environment variables only if they
haven't previously been set--otherwise, reused
the cached ones and the address.

This process (in `getIpcAddress`) would be vastly simpler if extensions
could read the environment variables that others provide, though there
may be security considerations since secrets are sometimes stashed
(though I could technically implement this today by manually creating
and terminal and running the appropriate `echo $FOO` command).

This seems to work fairly well in my testing. Fixes #88599.

/cc @isidorn @roblourens

### CODE REVIEW

**Comment by weinand** [comment]
When trying to use this PR I see the issue that js-debug does not automatically start debugging when setting "debug.node.useV3" to true.

Here are the steps:
- checkout the branch "connor4312/js-debug-auto-attach"
- run VS Code out of source on a trivial server.js project
- make sure to use this setting: `"debug.node.useV3": false`
- set a breakpoint in the server.js
- F1 > "auto attach": enable auto attach
- open the integrated terminal in VS Code (not the JavaScript terminal)
- from there run "node --inspect server.js"
Observe:
debugger starts and the breakpoint is hit

- stop debugging
- changing setting to `"debug.node.useV3": true`
- from the same integrated terminal run "node server.js"

Observe:
no debug session starts

The version of the Nightly js-debug is 2020.4.1517.

In addition I see this error:
```
[59161:0422/164507.675551:INFO:CONSOLE(104)] "%c[Extension Host] %cstack trace: Error: command 'extension.js-debug.setAutoAttachVariables' not found
    at CommandService._tryExecuteCommand (file:///Users/weinand/MS/Projects/vscode/out/vs/workbench/services/commands/common/commandService.js:71:43)
    at file:///Users/weinand/MS/Projects/vscode/out/vs/workbench/services/commands/common/commandService.js:65:51
    at processTicksAndRejections (internal/process/task_queues.js:85:5) %c(at /Users/weinand/MS/Projects/vscode/out/vs/workbench/services/extensions/node/extensionHostProce
```

**Comment by connor4312** [comment]
> The version of the Nightly js-debug is 2020.4.1517.

The changes into js-debug are still in a PR, I haven't merged/published them yet in case there were any structural changes coming out of this PR. Here's a vsix of that branch, you can install it manually to test this: https://memes.peet.io/img/debug-terminal-js-debug-nightly.vsix

**Comment by weinand** [comment]
Yep, with that vsix all my scenarios work just fine. Great stuff.

File: extensions/debug-auto-launch/src/extension.ts [hunk 1]
```diff
@@ -68,65 +88,163 @@ function toggleAutoAttachSetting() {
...
+/**
+ * Map of logic that happens when auto attach states are entered and exited.
+ * All state transitions are queued and run in order; promises are awaited.
+ */
+const transitions: { [S in State]: StateTransition<unknown> } = {
+	[State.Disabled]: {
+		async enter(context) {
+			statusItem?.hide();
+			// not strictly necessary, but clearing the cached state if autoattach is
+			// disabled provices an escape hatch if state gets corrupted somehow:
```
Comment (weinand) [inline extensions/debug-auto-launch/src/extension.ts:146]: provices -> provides

File: extensions/debug-auto-launch/src/extension.ts [hunk 2]
```diff
@@ -5,41 +5,61 @@
 
 import * as vscode from 'vscode';
 import * as nls from 'vscode-nls';
+import { createServer, Server } from 'net';
 
 const localize = nls.loadMessageBundle();
-const ON_TEXT = localize('status.text.auto.attach.on', "Auto Attach: On");
-const OFF_TEXT = localize('status.text.auto.attach.off', "Auto Attach: Off");
+const ON_TEXT = localize('status.text.auto.attach.on', 'Auto Attach: On');
+const OFF_TEXT = localize('status.text.auto.attach.off', 'Auto Attach: Off');
```
Comment (weinand) [inline extensions/debug-auto-launch/src/extension.ts:12]: with single ticks for the default localisation strings you are in the minority 😁 (1500 vs. 5000).

File: extensions/debug-auto-launch/src/extension.ts [hunk 3]
```diff
@@ -68,65 +88,163 @@ function toggleAutoAttachSetting() {
...
-				statusItem.show();
-				context.subscriptions.push(statusItem);
-			} else {
-				statusItem.show();
-			}
+/**
+ * Map of logic that happens when auto attach states are entered and exited.
+ * All state transitions are queued and run in order; promises are awaited.
+ */
+const transitions: { [S in State]: StateTransition<unknown> } = {
```
Comment (weinand) [inline extensions/debug-auto-launch/src/extension.ts:141]: very nice way to represent a state machine.

**Review by weinand — CHANGES_REQUESTED** [review]
code looks good to me. Just one typo to fix.

**Review by weinand — APPROVED** [review]

**Pull request merged**